File: backend/dev_date.py
"""Simulation de la date courante, réservée au développement.

En production, `get_current_date` renvoie toujours `date.today()` : ce module
n'a alors strictement aucun effet. Il n'est activé que si la variable
d'environnement DEV_MODE est positionnée côté backend (jamais en prod), et
seulement si le client envoie explicitement le header X-Dev-Date.

Ne modifie ni les données réelles, ni la date de création du programme, ni
les timestamps déjà enregistrés en base : ce module ne fait que fournir une
date "aujourd'hui" alternative aux endpoints qui en ont besoin.
"""

import os
from datetime import date, datetime
from typing import Optional

from fastapi import Header

DEV_MODE = os.environ.get("DEV_MODE", "").strip().lower() in ("1", "true", "yes")


def get_current_date(x_dev_date: Optional[str] = Header(default=None)) -> date:
    """Dépendance FastAPI : date "aujourd'hui" à utiliser par l'endpoint.

    Ignorée (retombe sur la vraie date système) sauf si DEV_MODE est actif
    côté backend ET que le header X-Dev-Date (format YYYY-MM-DD) est fourni.
    """
    if DEV_MODE and x_dev_date:
        try:
            return datetime.strptime(x_dev_date, "%Y-%m-%d").date()
        except ValueError:
            pass
    return date.today()

File: src/utils/devDate.ts
// Simulation de la date courante, pour tester le programme (semaine/jour, séance du jour,
// adaptation) sans attendre le temps réel.
//
// Ne modifie ni les données réelles ni la date système : fournit seulement une "date
// courante" alternative aux endroits qui en ont besoin (semaine/jour du programme, écran
// Aujourd'hui) et, via le header X-Dev-Date, au backend (voir backend/dev_date.py — lui-même
// gardé par sa propre variable d'environnement DEV_MODE côté serveur : sans elle, le backend
// ignore le header même si le panneau est visible côté frontend).
//
// Le panneau (DevDatePanel) est visible en build de développement (import.meta.env.DEV) OU,
// en production (ex: Vercel), une fois activé explicitement via l'URL ?devmode=1 — l'activation
// est alors mémorisée en localStorage. Visite ?devmode=0 pour la désactiver.

const STORAGE_KEY = 'level_dev_simulated_date';
const DEV_MODE_STORAGE_KEY = 'level_dev_mode_enabled';

function lireFlagDevModeDepuisUrl(): boolean | null {
  if (typeof window === 'undefined') return null;
  const valeur = new URLSearchParams(window.location.search).get('devmode');
  if (valeur === '1') return true;
  if (valeur === '0') return false;
  return null;
}

function devModeActivePersiste(): boolean {
  try {
    const depuisUrl = lireFlagDevModeDepuisUrl();
    if (depuisUrl !== null) {
      localStorage.setItem(DEV_MODE_STORAGE_KEY, depuisUrl ? '1' : '0');
      return depuisUrl;
    }
    return localStorage.getItem(DEV_MODE_STORAGE_KEY) === '1';
  } catch {
    return false;
  }
}

export const DEV_MODE_ENABLED = import.meta.env.DEV || devModeActivePersiste();

// Format YYYY-MM-DD, ou null si aucune simulation n'est active.
export function getDevSimulatedDate(): string | null {
  if (!DEV_MODE_ENABLED) return null;
  try {
    return localStorage.getItem(STORAGE_KEY);
  } catch {
    return null;
  }
}

export function setDevSimulatedDate(dateStr: string | null): void {
  if (!DEV_MODE_ENABLED) return;
  try {
    if (dateStr) {
      localStorage.setItem(STORAGE_KEY, dateStr);
    } else {
      localStorage.removeItem(STORAGE_KEY);
    }
  } catch {
    // localStorage indisponible (navigation privée, quota...) : simulation ignorée.
  }
}

// Remplace new Date() partout où le code doit connaître "aujourd'hui" côté frontend.
export function getNow(): Date {
  const override = getDevSimulatedDate();
  if (override) {
    const [y, m, d] = override.split('-').map(Number);
    return new Date(y, m - 1, d);
  }
  return new Date();
}
